File: androidApp/src/main/java/me/nathanfallet/bdeensisa/features/clubs/ClubViewModel.kt
package me.nathanfallet.bdeensisa.features.clubs

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.viewModelScope
import com.google.firebase.analytics.FirebaseAnalytics
import com.google.firebase.analytics.ktx.analytics
import com.google.firebase.analytics.ktx.logEvent
import com.google.firebase.ktx.Firebase
import kotlinx.coroutines.launch
import me.nathanfallet.bdeensisa.models.Club
import me.nathanfallet.bdeensisa.models.ClubMembership
import me.nathanfallet.bdeensisa.services.APIService

class ClubViewModel(
    application: Application,
    val club: Club
) : AndroidViewModel(application) {

    // Properties

    private val members = MutableLiveData<List<ClubMembership>>()

    // Getters

    fun getMembers(): LiveData<List<ClubMembership>> {
        return members
    }

    // Methods

    init {
        Firebase.analytics.logEvent(FirebaseAnalytics.Event.SCREEN_VIEW) {
            param(FirebaseAnalytics.Param.SCREEN_NAME, "club")
            param(FirebaseAnalytics.Param.SCREEN_CLASS, "ClubView")
        }

        fetchMembers()
    }

    fun fetchMembers() {
        viewModelScope.launch {
            try {
                APIService().getClubMembers(club.id).let {
                    members.postValue(it)
                }
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }

    fun join(token: String?) {
        if (token == null) {
            return
        }
        viewModelScope.launch {
            try {
                APIService().joinClub(token, club.id)
                fetchMembers()
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }

    fun leave(token: String?) {
        if (token == null) {
            return
        }
        viewModelScope.launch {
            try {
                APIService().leaveClub(token, club.id)
                fetchMembers()
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }

}

File: iosApp/iosApp/Features/Clubs/ClubViewModel.swift
//
//  ClubViewModel.swift
//  BDE
//
//  Created by Nathan Fallet on 13/02/2023.
//  Copyright © 2023 orgName. All rights reserved.
//

import Foundation
import shared

class ClubViewModel: ObservableObject {
    
    @Published var club: Club
    @Published var members = [ClubMembership]()
    
    init(club: Club) {
        self.club = club
    }
    
    func onAppear() {
        AnalyticsService.shared.log(.screenView(screenName: "club", screenClass: "ClubView"))
        
        fetchMembers()
    }
    
    func fetchMembers() {
        Task {
            let members = try await CacheService.shared.apiService().getClubMembers(id: club.id)
            DispatchQueue.main.async {
                self.members = members
            }
        }
    }
    
    func join(token: String?) {
        guard let token else {
            return
        }
        Task {
            try await CacheService.shared.apiService().joinClub(token: token, id: club.id)
            fetchMembers()
        }
    }
    
    func leave(token: String?) {
        guard let token else {
            return
        }
        Task {
            try await CacheService.shared.apiService().leaveClub(token: token, id: club.id)
            fetchMembers()
        }
    }
    
}
